Memory management for array allocation in nopython

sklam opened the pull request:
* Introduces a *numba runtime* or NRT
* New option to jit, njit ``nrt=True`` to enable NRT.
* support ``np.empty()`` in nopython+nrt
* return array, return slices of array in nopython+nrt
* return any array in nopython+nrt

ufechner7 commented:
How does this work? Do you use the memory allocator of Python, or a separate one? Do you use reference counting or garbage collection?

sklam commented:
This implementation uses a custom atomic reference counting.  The atomic operations are generated by LLVM for portability so you won't see it in the C code.  Cyclic reference is not yet handled.  Our first and major usecase is for arrays, which cannot have cyclic reference.  A separate reference count is maintained by NRT so that incref/decref do not need the GIL.  

At https://github.com/sklam/numba/commit/f75cdfd548474caf994da5dc15a5ef3d9c3fc287, the implementation is not completely threadsafe yet.  Internally, NRT maintains some data structures (the freelist for ``MemInfo``) that are subject race condition.  I am working on a threadsafe version by using atomic compare-and-swap.

gmarkall commented on numba/numpy_support.py:
```diff
@@ -123,6 +123,8 @@ def as_dtype(nbtype):
     if isinstance(nbtype, (types.CharSeq, types.UnicodeCharSeq)):
         letter = _as_dtype_letters[type(nbtype)]
         return numpy.dtype('%s%d' % (letter, nbtype.count))
+    if isinstance(nbtype, types.Record):
```
I don't think this is required - I think it would only be hit if you had a record within a record, which is not supported at the moment.

sklam replied:
This is not a required change.  Just that in some of my exploration, I wanted to call `as_dtype(my_numba_record)` for creating a a structure dtype.  Also, this will make `my_numba_record==from_dtype(as_dtype(my_numba_record))` work.

gmarkall commented on numba/tests/test_dyn_array.py:
```diff
@@ -0,0 +1,350 @@
+        np.testing.assert_equal(yy, np.ones(2, dtype=np.float32))
+        np.testing.assert_equal(out, np.ones(2, dtype=np.float32))
+
+    def test_array_pass_through(self):
```
Would it be meaningful to test that the refcount of the array is unchanged by calling the jitted function here?

sklam replied:
Yes, it will be good to check our refcount behavior.  Here, `y` is referenced by the numba IR which is kept alive by the dispatcher.

gmarkall commented on numba/lowering.py:
```diff
@@ -247,7 +257,13 @@ def lower_inst(self, inst):
-            pass
+            try:
+                # XXX: incorrect Del injection?
+                val = self.loadvar(inst.value)
```
IIUC, loadvar is creating a new load instruction that is not used - if this is the case, would calling getvar be enough to determine whether the `KeyError` is thrown?

sklam replied:
Yes, `getvar` will be enough.  The extra load wouldn't be a problem.  LLVM will optimize it out.  Ultimately, I want to improve the `Del` injection so I don't need the try-except here.

pitrou commented on numba/tests/test_dyn_array.py:
```diff
@@ -0,0 +1,362 @@
+        # out reference by cfunc
+        self.assertEqual(initrefct + 1, sys.getrefcount(y))
+
+        np.testing.assert_equal(y, out)
```
Is `out` the same object as `y`? Is it backed by the same memory? If yes, then test mutating one effects the other as well?

sklam replied:
`out` is not `y` in this code.  I haven't change the semantic of global array, yet.  I think it should be a separate pull request.

pitrou commented on numba/_arraystruct.h:
```diff
@@ -0,0 +1,21 @@
+ */
+
+typedef struct {
+    void     *meminfo;
```
Add a comment explaining where "meminfo" is described perhaps?

pitrou commented on numba/tests/test_dyn_array.py:
```diff
@@ -0,0 +1,362 @@
+
+
+
+nrtjit = njit(nrt=True)
```
I don't think it makes sense to expose "nrt" to the user. Numba should always allow memory allocation from CPU funcs, IMO.

sklam replied:
I think having `nrt=True` to be the default and allowing it to be disabled could be helpful for numba developers for testing and debugging.  We can make it `_nrt` to suggesting it is internal usage.  What do you think?

pitrou replied:
That would sound better, yes. Also, don't you want to add `nopython=True`?

pitrou commented on numba/runtime/_nrt_python.c:
```diff
@@ -0,0 +1,539 @@
+}
+
+static void
+NRT_adapt_buffer(Py_buffer *buf, arystruct_t *arystruct)
```
Does it have to be separate from the equivalent function in _helperlib.c?

sklam replied:
No, this is temporary.

Ultimately, I want 'nrt' be a shared library that has no python dependency for potential pycc use cases.  So that, pycc can compile standalone binaries that are linked nrt.  I see _nrt_python to be the bridge that provides the adaptors between nrt and python.

Thoughts?

pitrou replied:
Is pycc useful? We don't document it and we get no reports of its use in the wild.

seibert replied:
pycc is not currently useful (it's functionally was fundamentally broken in the Great Numba Refactor), but we do get requests from people for the ability to compile code such that the Numba module is no longer needed.

We have several things to resolve before that is possible, but I think we do want to plan for how nopython mode code can be made independent of Numba.

pitrou commented:
So, from a quick glance, the implementation adds a lot of complication that doesn't sound strictly necessary for a reference-counting scheme:
- the reference count is a separately allocated structure rather than an inline member
- it has its own freelist scheme
- it has a "deferred destructor" scheme (what is that, and why is it useful?)
- it has its own spinlock implementation (this sounds like a bad idea, IMHO)
- it maintains function pointers for the atomic increment and decrement functions

Ideally:
- the reference count field is all we have to add to existing structures, and it can be inlined (similar to PyObject_HEAD for CPython objects)
- the object type is known at compile time, so is the destructor
- there is no need to customize the increment and decrement functions at runtime
- if we want to handle lock-free structures, it sounds better to use an existing library :-)

sklam commented:
> So, from a quick glance, the implementation adds a lot of complication that doesn't sound strictly necessary for a reference-counting scheme:

> the reference count is a separately allocated structure rather than an inline member

In the emitted llvm, numba treats most data as immutable.  Structures are passed around within each function as LLVM structs.  To have a mutable field, we will need to make structures pass by reference (by pointer to the LLVM struct).  My implementation makes the MemInfo an opaque pointer that can be treated as an immutable field.  This avoids substantial changes to the compiler.

As to why we pass structures by value:
* it is faster (due to fewer indirection) on GPU architectures where memory bandwidth are scarce.
* easier ABI for functions since we only pass around non-aggregate, primitives types (and there are so few of them).

> it has its own freelist scheme

This is to avoid too much malloc/free calls.  But, I haven't proven it to be useful.  

> it has a "deferred destructor" scheme (what is that, and why is it useful?)

Memory backed by a PyObject requires the GIL for Py_DecRef, which is called once the NRT refct drops to zero.  I am considering if it would be useful to allow the destructor be deferred so that we can call perform multiple Py_DecRef in a GIL acquire+release.  Otherwise, we are calling one Py_DecRef per GIL acquire+release.

> it has its own spinlock implementation (this sounds like a bad idea, IMHO)

I am using LLVM to generate the atomic operations for portability reason.  That way I don't have to change anything when we port to a new architecture.  Of course, we can always substitute it with a portable atomic+lock library.

> it maintains function pointers for the atomic increment and decrement functions
see comment above.

> Ideally:

> the reference count field is all we have to add to existing structures, and it can be inlined (similar to PyObject_HEAD for CPython objects)

see comment above

> the object type is known at compile time, so is the destructor

We only differentiate between NRT allocated vs Python allocated at this time.  The compiler needs to insert type-specific destructor for types containing other allocated objects.  But, a function does not know who allocated the memory.

> there is no need to customize the increment and decrement functions at runtime
> if we want to handle lock-free structures, it sounds better to use an existing library :-)

Can you suggest a library?

pitrou commented:
> In the emitted llvm, numba treats most data as immutable. Structures are passed around within each function as LLVM structs

Oh, yes, sorry for that. I always forget it (my mistake for often dealing with PyObject pointers :-)).

That said, the MemInfo could still be stack-allocated in the Numba function (a bit like generators), rather than heap-allocated. I wonder if that would be complicated to implement...

> Memory backed by a PyObject requires the GIL for Py_DecRef, which is called once the NRT refct drops to zero.

I don't understand how a Numba-compiled function can ever call Py_DecRef in nopython mode. That means the object was passed as parameter to the function, but then we shouldn't need to memory-manage it ourselves...

> I am using LLVM to generate the atomic operations for portability reason

I am talking about the lock / unlock primitives in the NRT lib. AFAIU, tuning a spinlock is delicate (you want to yield after a certain # of failures, otherwise you may waste a whole timeslice waiting for the unlock).

By the way, if the freelist is just a single-linked list, perhaps you can just atomically replace the list head without any locking? Something like:

```
list *newhead = malloc(...);
do {
    list *head = global_head;   /* volatile */
    newhead->next = head;
} while (!atomic_cas(&global_head, head, newhead));
```

> Can you suggest a library?

I'll search for one if this is really necessary. But I wonder if we could simply not have any spinlock at all.

sklam commented:
> I don't understand how a Numba-compiled function can ever call Py_DecRef in nopython mode. That means the object was passed as parameter to the function, but then we shouldn't need to memory-manage it ourselves...

It can happen with container types once we have them. 

> By the way, if the freelist is just a single-linked list, perhaps you can just atomically replace the list head without any locking? Something like:

Yes, I will try that.

pitrou commented:
> It can happen with container types once we have them. 

IMO that's a bit prematurate :-) Also, if we want a specialized representation of container types, not a bunch of PyObject pointers, we wouldn't share the representation with the CPython one...

sklam commented:
It is for the case when nrt is borrowing memory owned by a pyobject.

Let's say we have:

```
class Foo(object):
    def __init__(self, array):
        self.array = array
```

If we don't want to copy the data of array (a numpy.ndarray), MemInfo needs to keep a reference to the original object.  Decref on MemInfo may trigger a Py_Decref on the original pyobject.

sklam commented:
> IMO that's a bit prematurate :-) 

I agree.  It is not fully developed.  It wasn't turn on by default.  I just wanted to make sure it is possible.  
It probably need to trigger the cleanup once awhile to avoid accumulating too many deferred dtors.

seibert commented:
Note that the immediate followup to this work is going to be working on JIT classes, so I don't think support for retention of arrays inside a container is that premature.  We certainly want to think through that case before we commit to this approach to memory management.

gmarkall commented on numba/runtime/nrt.py:
```diff
@@ -0,0 +1,52 @@
+        mi = _nrt.meminfo_new(data, pyobj)
+        return MemInfo(mi)
+
+    def meminfo_alloc(self, size, safe=False):
```
`safe` here appears to mean "allocated memory is zero-filled" (if I've understood correctly) - perhaps this should be documented or renamed to something more descriptive? (or is "safe" == "zero-filled" a convention?)

sklam replied:
I added this during debugging.  I wanted it to mean anything that makes allocation safer.  All I could think of is zero-filling after malloc() and before free().  It helped track down premature deallocation.  I will document what it does.  We can enhance it to do more.  I don't expect it to be enabled for optimized run.

pitrou commented on numba/lowering.py:
```diff
@@ -185,8 +185,18 @@ def lower_inst(self, inst):
             self.context.debug_print(self.builder, str(inst))
         if isinstance(inst, ir.Assign):
             ty = self.typeof(inst.target.name)
+            try:
```
Should this additional code be put in storevar() instead?

pitrou commented on numba/pythonapi.py:
```diff
@@ -1201,6 +1217,31 @@ def from_native_generator(self, val, typ, env=None):
+                                                           self.voidptr),
+                                      ndim, typenum])
+
+    def nrt_array_adaptor(self, ary, ptr):
```
The difference between "nrt_array_adaptor" and "nrt_adapt_native_array" is not obvious at first sight :-)

pitrou commented on numba/runtime/_nrt_python.c:
```diff
@@ -0,0 +1,539 @@
+static
+PyObject*
+memsys_set_atomic_inc_dec(PyObject *self, PyObject *args) {
+    unsigned PY_LONG_LONG addr_inc, addr_dec;
```
I'm not sure PY_LONG_LONG is always enough for a pointer, but it probably is on common architectures... Otherwise, the right way would be PyLong_AsVoidPtr().

pitrou commented on numba/runtime/_nrt_python.c:
```diff
@@ -0,0 +1,539 @@
+    PyGILState_STATE gstate;
+    PyObject *ownerobj = info;
+
+    gstate = PyGILState_Ensure();   /* ensure the GIL */
```
If the destructors are always "deferred", you shouldn't need this (i.e. the destructor is called either from the wrapper or Python itself).

sklam replied:
We do not always defer dtors.

pitrou commented on numba/runtime/_nrt_python.c:
```diff
@@ -0,0 +1,539 @@
+meminfo_alloc(PyObject *self, PyObject *args) {
+    MemInfo *mi;
+    Py_ssize_t size;
+    if (!PyArg_ParseTuple(args, "K", &size)) {
```
The code for Py_ssize_t is "n", not "K".

pitrou commented on numba/runtime/_nrt_python.c:
```diff
@@ -0,0 +1,539 @@
+}
+
+#if (PY_MAJOR_VERSION < 3)
+static PyBufferProcs MemInfo_bufferProcs = {MemInfo_rdwrbufferproc,
```
Instead of that, you can implement the new buffer protocol on Python 2. Enable the Py_TPFLAGS_HAVE_NEWBUFFER and see here for an example:
https://hg.python.org/cpython/file/4bf210f59ac6/Objects/bytearrayobject.c#l2800

sklam replied:
After some experiment, Py_TPFLAGS_HAVE_NEWBUFFER does not simplify much because numpy and possibly many other packages do not use the check for it.

pitrou commented on numba/runtime/_nrt_python.c:
```diff
@@ -0,0 +1,539 @@
+
+static
+PyObject*
+MemInfo_release_defer(MemInfoObject *self) {
```
Do we really need a separate method here? This is just for testing, right?

sklam replied:
It is not used in the current code.  Maybe I should get rid of it.

pitrou replied:
Yes, I think getting rid of unused APIs will make the whole thing more palatable :-)

pitrou commented on numba/runtime/_nrt_python.c:
```diff
@@ -0,0 +1,539 @@
+     "Decrement the reference count but defer the destructor regardless"
+     "of the value in defer attribute"
+    },
+    {"set_defer", (PyCFunction)MemInfo_set_defer, METH_VARARGS,
```
Just for the record, set_defer / get_defer could have been a "getset" instead (there's an example in numba/npyufunc/_internal.c).

pitrou commented on numba/runtime/atomicops.py:
```diff
@@ -0,0 +1,211 @@
+    # Implement LLVM module with atomic ops
+    mod = llvm.parse_assembly(str(define_atomic_ops()))
+
+    # Create a target just for this module
```
Can't you reuse the codegen infrastructure here?

pitrou commented on numba/runtime/nrt.c:
```diff
@@ -0,0 +1,271 @@
+MemInfo *nrt_pop_meminfo_list(MemInfo * volatile *list) {
+    MemInfo *old, *repl, *head;
+
+    head = *list;     /* get the current head */
```
You certainly want this inside the loop (since another thread may change the head pointer).

sklam replied:
atomic_cas also atomically loads. see https://github.com/numba/numba/pull/1077/files#diff-b9938948c081b5f404bb98f9dffa348aR54 and http://llvm.org/docs/LangRef.html#id185

pitrou commented on numba/runtime/nrt.py:
```diff
@@ -0,0 +1,68 @@
+        _nrt.memsys_set_atomic_cas(self._ptr_cas)
+
+    def __del__(self):
+        _nrt.memsys_shutdown()
```
Since the "memsys" is a singleton, the Runtime object can only be used once, right? If so, I think it should be instantiated as a global in this module. Otherwise, it's too easy to make a mistake.

pitrou commented on numba/runtime/nrt.c:
```diff
@@ -0,0 +1,271 @@
+                         void *dtor_info)
+{
+    MemInfo * mi = NRT_MemSys_pop_meminfo();
+    mi->payload.refct = 0;
```
This feels a bit weird. You should start with a refcount of 1, no?

pitrou commented on numba/runtime/atomicops.py:
```diff
@@ -0,0 +1,211 @@
+    """Define a llvm function for atomic increment/decrement to the given module
+    Argument ``op`` is the operation "add"/"sub".  Argument ``ordering`` is
+    the memory ordering.
+    """
```
Please mention in the docstring that the function returns the new value (while LLVM's atomic_rmw returns the original value).

pitrou commented on numba/runtime/atomicops.py:
```diff
@@ -0,0 +1,211 @@
+    builder = ir.IRBuilder(bb)
+    ONE = ir.Constant(_word_type, 1)
+    builder.atomic_rmw(op, ptr, ONE, ordering=ordering)
+    res = builder.load(ptr)
```
I think there is a race condition. If another thread runs in-between, then the two threads will see a resulting refcount of 0, so the destructor will be run twice. You want to reapply `op` to the result of the atomic_rmw instruction, instead (or you can change the decref test to check for 1, for example).

sklam replied:
Good catch.

pitrou commented on numba/runtime/atomicops.py:
```diff
@@ -0,0 +1,211 @@
+_regex_bb = re.compile(r'[-a-zA-Z$._][-a-zA-Z$._0-9]*:')
+
+
+def remove_redundant_nrt_refct(ir_module):
```
Actually, this is called with a LL module, not an IR module, right?

pitrou commented on numba/targets/arrayobj.py:
```diff
@@ -1238,3 +1224,54 @@ def iternext_numpy_ndindex(context, builder, sig, args, result):
+
+@builtin
+@implement(numpy.empty, types.Kind(types.Integer))
+@implement(numpy.empty, types.Kind(types.Tuple))
```
By the way, you can now use types.BaseTuple to match both Tuple and UniTuple.

pitrou commented on numba/tests/test_dyn_array.py:
```diff
@@ -0,0 +1,362 @@
+
+        cfunc = nrtjit(pyfunc)
+        size = 10**5
+        arr = np.random.randint(0, 1, size)
```
This creates a zero-filled array.

pitrou commented on numba/tests/test_dyn_array.py:
```diff
@@ -0,0 +1,362 @@
+            inputs.append(arr)
+            outputs.append(out)
+
+        # Launch worker threads
```
I'm not sure this tests much. The threads will be serialized, since you didn't add `nogil=True`. Furthermore, they all operate on distinct arrays, so there's not much chance for contention (is there?).

sklam replied:
> Furthermore, they all operate on distinct arrays, so there's not much chance for contention (is there?).
I was testing the CAS for the freelist.  I should add one for the refct as well.

sklam replied:
This first one is actually not very interesting.  I will discard it.

pitrou commented on numba/tests/test_dyn_array.py:
```diff
@@ -0,0 +1,362 @@
+            workers.append(thread)
+            outputs.append(out)
+
+        # Launch worker threads
```
Similarly, I'm not sure what you're testing here. As long as the threads are serialized and don't access the same objects, there's not much to test for :-) (the freelist perhaps?)

pitrou commented on numba/tests/test_nrt.py:
```diff
@@ -0,0 +1,174 @@
+        type(self).alive -= 1
+
+
+class TestNrtMemInfo(unittest.TestCase):
```
If you don't want a test failure to ruin all subsequent tests, you could add `Dummy.alive = 0` in a setUp() method.

sklam replied:
Applied at d3ed346

pitrou commented on numba/typing/npydecl.py:
```diff
@@ -318,6 +318,29 @@ def generic(self, args, kws):
+        dtype = types.double
+        if len(args) >= 2:
+            npy_dtype = args[1]
+            dtype = from_dtype(numpy.dtype(npy_dtype.template.key))
```
Is there a reason for the complication? Also, I'm wondering: does it work with a structured dtype?

sklam replied:
I will add comments.  It's because things like numpy.int32 is not a dtype but a dtype constructor.

sklam commented:
Needs to check a reference count problem

sklam commented:
Important changes since last review:
* keyword arg `nrt" renamed to "_nrt" and defaults to False
* Buffer object also use NRT if it has a owner object
* Refct initialize to 0 for easier code generation (see https://github.com/sklam/numba/commit/e4e1270aa24635678ba929b7093634454833e44b)
